### PR TITLE
Use CDI API for clone authorization

### LINKED ISSUE #9574: Can't create VM from Snapshot via CDI due to missing "snapshot" in spec
**What happened**:
It is not possible to create a new VirtualMachine from containerized-data-importer with source `snapshot` caused by `unknown field "spec.dataVolumeTemplates[0].spec.source.snapshot"` error.

See kubevirt API reference:
http://kubevirt.io/api-reference/v0.59.0/definitions.html#_v1beta1_datavolumesource

See kubevirt CDI API reference:
https://kubevirt.io/cdi-api-reference/v1.56.0/definitions.html#_v1beta1_datavolumesource

**What you expected to happen**:
VirtualMachine from Snapshot is created

**How to reproduce it (as minimally and precisely as possible)**:
```
  dataVolumeTemplates:
    - metadata:
        name: "release-name"
      spec:
        pvc:
          accessModes: [ReadWriteOnce]
          resources:
            requests:
              storage: "10Gi"
        source:
          snapshot:
            name: "snap"
            namespace: "default"
        contentType: "kubevirt"
```

**Environment**:
- KubeVirt version (use `virtctl version`): `v0.59.0`
- Kubernetes version (use `kubectl version`): `v1.25.6`
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: `bare-metal`
- OS (e.g. from /etc/os-release): `Debian`
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by akalenyu**
**What this PR does / why we need it**:
CDI API can now hide away the authorization pieces (DataSources, clone types) behind dv.Authorize() which is now useable,
plus, this will also take snapshot clone auth into account.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #9574 

**Special notes for your reviewer**:

**Release note**:

```release-note
Use auth API for DataVolumes, stop importing kubevirt.io/containerized-data-importer
```

### CODE REVIEW

**Comment by akalenyu** [comment]
> /retest-required

Just rebased, no conflicts

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go [hunk 1]
```diff
@@ -257,49 +272,34 @@ func (admitter *VMsAdmitter) authorizeVirtualMachineSpec(ar *admissionv1.Admissi
...
 		}
 
-		allowed, message, err := admitter.cloneAuthFunc(cloneSource.Namespace, cloneSource.Name, targetNamespace, serviceAccount)
-		if err != nil {
+		proxy := &authProxy{client: admitter.VirtClient, dataSourceInformer: admitter.DataSourceInformer}
+		dv := &cdiv1.DataVolume{
+			ObjectMeta: dataVolume.ObjectMeta,
+			Spec:       dataVolume.Spec,
+		}
+		dv.Namespace = targetNamespace
```
Comment (mhenriks) [inline pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go:300]: Oh man, the the fact that you have to create a tmp DataVolume and explicitly set the namespace bums me out pretty bad.  Nothing we can do now but wish we thought of that earlier.

File: pkg/virt-controller/watch/vm.go [hunk 2]
```diff
@@ -179,14 +177,22 @@ func NewVMController(vmiInformer cache.SharedIndexInformer,
...
+func (p *authProxy) CreateSar(sar *authv1.SubjectAccessReview) (*authv1.SubjectAccessReview, error) {
 	return p.client.AuthorizationV1().SubjectAccessReviews().Create(context.Background(), sar, v1.CreateOptions{})
 }
 
+func (p *authProxy) GetNamespace(name string) (*k8score.Namespace, error) {
+	return p.client.CoreV1().Namespaces().Get(context.TODO(), name, v1.GetOptions{})
+}
+
+func (p *authProxy) GetDataSource(namespace, name string) (*cdiv1.DataSource, error) {
+	return p.client.CdiClient().CdiV1beta1().DataSources(namespace).Get(context.TODO(), name, v1.GetOptions{})
```
Comment (mhenriks) [inline pkg/virt-controller/watch/vm.go:193]: should use informers for namespace and datasource

File: pkg/virt-operator/resource/generate/rbac/apiserver.go [hunk 3]
```diff
@@ -218,6 +218,17 @@ func newApiServerClusterRole() *rbacv1.ClusterRole {
...
 			},
+			{
+				APIGroups: []string{
+					"",
+				},
+				Resources: []string{
+					"namespaces",
+				},
+				Verbs: []string{
+					"get",
```
Comment (mhenriks) [inline pkg/virt-operator/resource/generate/rbac/apiserver.go:229]: this will have to be updated when using informers

**Review by mhenriks — COMMENTED** [review]
Looks good just have to use informers

File: pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go [hunk 4]
```diff
@@ -62,24 +60,41 @@ type VMsAdmitter struct {
...
 
-func (p *sarProxy) Create(sar *authv1.SubjectAccessReview) (*authv1.SubjectAccessReview, error) {
+func (p *authProxy) CreateSar(sar *authv1.SubjectAccessReview) (*authv1.SubjectAccessReview, error) {
 	return p.client.AuthorizationV1().SubjectAccessReviews().Create(context.Background(), sar, metav1.CreateOptions{})
 }
 
-func NewVMsAdmitter(clusterConfig *virtconfig.ClusterConfig, client kubecli.KubevirtClient, informers *webhooks.Informers) *VMsAdmitter {
-	proxy := &sarProxy{client: client}
+func (p *authProxy) GetNamespace(name string) (*corev1.Namespace, error) {
+	return p.client.CoreV1().Namespaces().Get(context.TODO(), name, metav1.GetOptions{})
```
[the remark this answers is not in the document]
Comment (akalenyu) [reply pkg/virt-api/webhooks/validating-webhook/admitters/vms-admitter.go]: done

File: pkg/virt-operator/resource/generate/rbac/apiserver.go (hunk 3 above)
Comment (akalenyu) [reply pkg/virt-operator/resource/generate/rbac/apiserver.go:229]: yup, thx

File: pkg/virt-controller/watch/vm.go (hunk 2 above)
Comment (akalenyu) [reply pkg/virt-controller/watch/vm.go]: done

File: pkg/virt-controller/watch/vm.go [hunk 5]
```diff
@@ -179,20 +181,49 @@ func NewVMController(vmiInformer cache.SharedIndexInformer,
...
 
+func (p *authProxy) GetNamespace(name string) (*k8score.Namespace, error) {
+	obj, exists, err := p.namespaceStore.GetByKey(name)
+	if err != nil {
+		return nil, err
+	} else if !exists {
+		return nil, fmt.Errorf("namespace %s does not exist", name)
+	}
+
+	ns := obj.(*k8score.Namespace)
```
Comment (mhenriks) [inline pkg/virt-controller/watch/vm.go:202]: When dealing with informers, I think it is good practice to return/work with a deep copy of the object in the cache.

Comment (akalenyu) [reply pkg/virt-controller/watch/vm.go]: makes sense, done

**Pull request merged**